---
output:
  html_document:
    includes:
      before_body: include/nav.html
      after_body: include/nothing.html
---


# Statistical Methods for High Dimensional Biology
<span style="color: red">**Next deadline: Finalized Project Proposal due - Feb. 13th, 2019**</span>

Please complete <span style="color: blue">[this survey](https://docs.google.com/forms/d/e/1FAIpQLSd8WlRLiBRaNrNk5PEY87nLkEPjZ60BcdTMUU0JAQRyMkkNsg/viewform?usp=sf_link)</span> as soon as you see this post so we can set up a repo for you! :) 

### UBC STAT 540

STAT 540 is a 3 credit course with a mandatory computing seminar.

Cross-listed as STAT 540, BIOF 540, GSAT 540

### Time and Location

02 January 2019 - 03 April 2019

#### Lecture (Sec 201)

**Time**: Mon Wed 9:30 - 11am

**Location**: ESB 2012

#### Seminar / computing lab (S2A) -- REGISTRATION IS REQUIRED!

**Time**: every Wednesday 11am - 1pm. The TAs will remain in the lab until 1pm for general office hours.

**Location**: Hugh Dempster Pavillion 201

### Learning objectives:

  * Explore and visualize genomics data
  * Apply tailored statitical methods for high dimensional biology data
  * Make your work reproducible, reusable, and shareable
  * Work with real data in a collaborative model

### Selected topics

  * Exploration, visualization, and quality assessment of high dimensional genomics data
  * Large scale statistical inference
  * Analysis of microarray, RNASeq, and epigenetics data
  * Principal component analysis (PCA)
  * Cluster analysis
  * Cross validation
  * Model selection for classification and linear regression models
  * Gene function and gene set analysis
  * Resampling and the bootstrap

### More info?

Use the navigation bar above!

The course organization on GitHub: <https://github.com/STAT540-UBC>  

Repo that creates this website: <https://github.com/STAT540-UBC/STAT540-UBC.github.io>

Repo that supports the Discussion forum: <https://github.com/STAT540-UBC/Discussion>

### Major deadlines
```{r, echo=FALSE}
suppressPackageStartupMessages(library(tidyverse))
assn <- read.csv("course-admin/assn_2019.csv")
assn <- assn %>% 
  select(content,start, Category) %>%
  rename(Assignment = content, `Due Date` = start) %>%
  mutate(`Due Date` = as.Date(`Due Date`, "%Y-%m-%d")) %>%
  arrange(`Due Date`) %>% 
  mutate(`Due Date` = format(`Due Date`, "%a %d %B %Y"))
knitr::kable(assn)
```

### Assignment Timeline

```{r,echo=FALSE}
suppressPackageStartupMessages(library(timevis))

assignment_data <- read.csv("course-admin/assn_2019.csv",as.is = T,sep = ",")
assignment_data$start <- paste0(assignment_data$start, " 23:59:00")
timevis(assignment_data)
```

