---
title: "Microbiome <br>data analysis"
author: "Kim Dill-McFarland, PhD (kadm@mail.ubc.ca)"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  ioslides_presentation:
    css: default.css
urlcolor: blue
editor_options: 
  chunk_output_type: console
---
```{r packages, echo=FALSE, message=FALSE, warning=FALSE}
# Data manipulation and plotting
library(tidyverse)
library(plyr)

# Phylogenetic analyses
library(phyloseq)
# Beta-diversity and other microbial ecology analyses
library(vegan)
# Differential abundance analysis
library(DESeq2)
```

```{r data, echo=FALSE, message=FALSE, warning=FALSE}
# Load phyloseq object of Saanich OTU counts, metadata, and taxonomy
load("data/Saanich_phyloseq.RData")

# Data can be downloaded with
### library(repmis)
### source_data("https://github.com/EDUCE-UBC/workshop_data/blob/master/Saanich_phyloseq.RData?raw=true")

# Add oxygen level variable to metadata
sample_data(physeq)$O2_group <- c("high", "present", "present", "present", "absent", "absent", "absent")
# Set seed for reproducibility
set.seed(4832)
```

## 
<div style="float: left; width: 50%;">

<br>

<br>

![](images/EDUCE_Globe.png)

</div>

<div style="float: left; width: 50%;">

<br>

<br>

<font size="6"><span style="color:teal">**E**</span>xperiential  
<span style="color:teal">**D**</span>ata science for  
<span style="color:teal">**U**</span>ndergraduate  
<span style="color:teal">**C**</span>ross-disciplinary  
<span style="color:teal">**E**</span>ducation</font>

<br>
<br>
<br>

</div>

If you want to learn more, email info.educe@ubc.ca or checkout our website at https://educe-ubc.github.io/

## Learning objectives

* Define microbiome and microbiota
* Describe computational and statistical challenges in microbiome research
* Assess microbiome data using beta-diversity

## {.flexbox .vcenter}

<font size="8">What is a microbiome?</font>

## Term clarification

Microbial community present in a particular environment

<br>

>* **Microbiome**: all combined genetic material of the microorganisms in a particular environment

<br>

>* **Microbiota**: the microorganisms in a particular environment

## Studying microbial communities - Past

![](images/sampling_pipeline_culture.png)

## Studying microbial communities - Now

![](images/sampling_pipeline_seq.png)

## Moore's Law
<center>
![](images/Moores_law.png)
</center>

## Waypoints in history

**2000**: Lynx Therapeutics massively parallel signature sequencing (MPSS)

**2004**: Roche 454 parallel pyrosequencing

**2007**: Illumina buys Solexa which bought Lynx

**2015**: Roche shuts down 454 sequencing

## The "microbiome" is born
<center>
![](images/word_usage.png)
</center>

## Let’s talk numbers
<div style="float: left; width: 50%;">
Sanger

* 150,000 bp / day
* Up to 1,000 bp long

Roche GX-FLX 454

* Up to 4.5 million bp / day
* Up to 700 bp long

Illumina MiSeq or HiSeq

* Up to 1.2 trillion bp / day
* Up to 300 bp long
</div>

<div style="float: right; width: 50%;">
>* __++ Number__

>* __--- Length__

</div>

## Challenges with microbiome data

1. Big data, high complexity

>* Solution: obtain high coverage and assemble into longer sequences
>* __Solution: targeted sequencing of specific gene(s)__

## Identity genes

* Requirements:

  >* All species (within a group) must have the gene
  >* The gene must contain enough differences to tell related species apart
  >* The gene must have both conserved and variable regions

## Bacterial 16S rRNA gene

<center>
![](images/16S.png)
</center>

## 16S pipeline

![](images/sampling_pipeline_seq.png)

## Raw 16S data

DNA sequences separated by observation

```{bash echo=FALSE}
head data/example.fasta
```

## 16S data cleaning

* Remove sequencing errors
* Align to a database
* Cluster very similar sequences (assumed error)
* Remove chimeras
* Classify to a database**

## Challenges with microbiome data

1. Big data, high complexity
2. Microbial species definition

>* Solution: Operational taxonomic units (OTUs)
>* Solution: Amplicon sequence varients (ASVs)

## Operational taxonomic units (OTUs)

![](images/OTU2.png)

## Operational taxonomic units (OTUs)

![](images/OTU.png)

## Amplicon sequence variants (ASVs)

![](images/ASV.png)

## OTU vs. ASV
<div style="float: left; width: 50%;">

**OTU**

* More established in the literature (2000-)
* Keeps more data 
* Uses representative sequence of each OTU to determine taxonomy
* Various levels (_e.g._ 97% = species)

</div>

<div style="float: right; width: 50%;">

**ASV**

* Less established and newer (2017-)
* Discards more data 
* Treats each ASV as a “species”

</div>

## 16S OTU data

```{r echo=FALSE, message=FALSE, warning=FALSE}
# Extract OTU table
OTU.count <- as.data.frame(otu_table(physeq))
# View OTU table
OTU.count[,1:7]
```

## Challenges with microbiome data

1. Big data, high complexity
2. Microbial species definition
3. Unequal sequencing coverage

## Challenges with microbiome data

1. Big data, high complexity
2. Microbial species definition
3. Unequal sequencing coverage

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.height=2.5}
# Sum total sequences per sample
OTU.count %>% 
  
ggplot(aes(x=rownames(.), y=rowSums(.))) +
  geom_col() +
  #Beautify
  labs(x="", y="Total sequences") +
  theme_classic() +
  theme(text = element_text(size=18),
        axis.text.x=element_blank())
```

## Challenges with microbiome data

1. Big data, high complexity
2. Microbial species definition
3. Unequal sequencing coverage

<br>

>* __Solution: Percent relative abundance__
>* Solution: Subsampling / rarefying
>* Solution: Variant stabilization
>* Solution: Inference of missing data

## 16S relative abundance OTU data

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}
# Calculate percent relative abundance
phy.perc <- transform_sample_counts(physeq, function(x) 100 * x/sum(x))

# Plot Phyla level groups as percent relative abundance
plot_bar(phy.perc) + 
  geom_bar(aes(fill=Phylum), stat="identity") +
  # Beautify
  theme_classic() +
  theme(text = element_text(size=18), legend.position = "none",
        axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(y="Percent relative abundance", x="")
```

## Challenges with microbiome data

1. Big data, high complexity
2. Microbial species definition
3. Unequal sequencing coverage
4. Skewed, sparse, dependent data

## Skewed data

```{r echo=FALSE, message=FALSE, warning=FALSE}
# Extract OTU table
OTU.perc <- as.data.frame(otu_table(phy.perc))

# Plot OTU histogram within 1 sample
OTU.perc %>% 
  rownames_to_column() %>% 
  # Select 10 m sample
  filter(rowname=="Saanich_010") %>% 
  # Gather OTUs
  gather(key=OTU, value=abund, -rowname) %>% 
  # Remove very low abundance
  filter(abund>0.01) %>% 
  
ggplot() +
  # Order bars by abundance
  geom_col(aes(x=reorder(OTU, -abund), y=abund)) +
  # Beautify
  labs(x="OTU", y="Percent relative abundance") +
  theme_classic() +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.line.x=element_blank(),
        text = element_text(size=18))
```

## Sparse data

```{r echo=FALSE, message=FALSE, warning=FALSE}
# View percent data
OTU.perc[,1:5]
```

>* 7 samples, 3,754 OTUs

>* 26,278 OTU counts

>* __75% zeros__

## Sparse data

```{r echo=FALSE, message=FALSE, warning=FALSE}
# Scroll to less abundant OTUs
OTU.perc[,1001:1005]
```

>* Rely on statistical analyses
>* Still searching for solutions

## Pipeline recap

1. Isolate DNA
2. Amplify and sequence a piece of the bacterial 16S
3. Remove poor quality and erroneous sequences
4. Cluster sequences into OTUs (*i.e.* species)
5. Normalize to relative abundance

## {.flexbox .vcenter}

<font size="8">
What next?
</font> 

## More on our example dataset {.flexbox .vcenter}

![](images/Saanich.png)

## Saanich Inlet sampling

![](images/saanich_photos.png)

## Saanich Inlet sampling

<iframe src="https://www.youtube.com/embed/XYQSm2Me86I" encrypted-media; gyroscope; picture-in-picture"></iframe>

https://www.youtube.com/embed/XYQSm2Me86I

## The Saanich Inlet model system

Oxygen stratification

![](images/oxygen_mean.png)

## The Saanich Inlet model system

Annual cycles of stratification and renewal

![](images/oxygen_timeseries.png)

## Oxygen in the ocean

<center>
![](images/ocean_o2.png)
</center>

## Global ocean deoxygenation
Year (expected) deoxygenation will be detectable

<center>
![](images/ocean_o2_time.png)
</center>

## Global ocean deoxygenation

* 20%+ reduction in O~2~ over past 50 years at several monitoring stations worldwide

* Widespread deoxygenation by 2030-2040

## Consequences of deoxygenation?

* Unknown how deoxygenation will impact the ocean ecosystem including:

>* Global nutrient cycles
* Primary production like photosynthesis
* Fish populations
* Marine mammal survival
* …

## {.flexbox .vcenter}

<font size="8">
How is global ocean deoxygenation impacting marine microbial communities?
</font>

## Beta-diversity

* Between sample diversity

* Measure of overall microbiome differences

<br>

>* Can take into account:

>* Presence/absence of OTUs
* Relative abundance of OTUs
* Taxonomic relatedness of OTUs

## Calculate beta-diversity
Bray-Curtis: Presence/absence and abundance but no taxonomy

```{r message=FALSE, warning=FALSE, echo=FALSE}
# Normalize data
## You could run these analyses on the percent relative abundance or normalized counts (as done here)
phy.norm <- rarefy_even_depth(physeq, sample.size = 10000)
# Extract OTU table
OTU.norm <- as.data.frame(otu_table(phy.norm))
```

```{r message=FALSE, warning=FALSE}
# Calculate Bray-Curtis distance
vegdist(OTU.norm, method="bray")
```

## Visualize beta-diversity
Non-metric multidimensional scaling (nMDS)

* Reduce all pairwise comparisons to distances within 2-dimensional plane

```{r warning=FALSE}
# Calculate nMDS from Bray-Curtis distance
## Conform to 2 axes (k) and 100 max iterations
nMDS <- metaMDS(OTU.norm, k=2, trymax=100, distance="bray")
```

## Visualize beta-diversity

```{r echo=FALSE}
# View nMDS fit results
nMDS
```

## Visualize beta-diversity

```{r}
# Extract nMDS xy coordinates
nmds.xy <- as.data.frame(scores(nMDS))
nmds.xy
```

## Visualize beta-diversity
```{r echo=FALSE, message=FALSE}
nmds.data <- 
  # Extract metadata
  as.data.frame(sample_data(phy.norm)) %>% 
  # Combine nMDS xy with metadata
  cbind(., nmds.xy) %>% 
  # Move rownames
  rownames_to_column()
  
# Plot basic nmds
ggplot(nmds.data, aes(x=NMDS1, y=NMDS2)) +
  geom_point(size=5) +
  theme_classic() +
  theme(text = element_text(size=18))
```

## Add metadata
```{r echo=FALSE, message=FALSE}
# Color by sample depth
ggplot(nmds.data, aes(x=NMDS1, y=NMDS2)) +
  geom_point(size=5, aes(color=as.factor(Depth_m))) +
  theme_classic() +
  theme(text = element_text(size=18)) +
  guides(color=guide_legend(title="Depth (m)"))
```

## Add metadata
```{r echo=FALSE, message=FALSE}
# Color by oxygen concentration
ggplot(nmds.data, aes(x=NMDS1, y=NMDS2)) +
  geom_point(size=5, aes(color=O2_uM)) +
  theme_classic() +
  theme(text = element_text(size=18)) +
  guides(color=guide_legend(title="Oxygen (uM)"))
```

## Add metadata
```{r echo=FALSE, message=FALSE}
# Color by oxygen levels
ggplot(nmds.data, aes(x=NMDS1, y=NMDS2)) +
  geom_point(size=5, aes(color=O2_group)) +
  theme_classic() +
  theme(text = element_text(size=18)) +
  guides(color=guide_legend(title="Oxygen"))
```

## A more exciting example

<center>

![](images/calf_nmds.png) 

Dill-McFarland *et al* 2017 Sci Rep

</center>

## Drawing conclusions from nMDS

* Clear separation =/= significant

* Lack of clear separation =/= not significant

<br>

>* __A sometimes useful way to visualize beta-diversity__

## Statistically assess beta-diversity
Permutational ANOVA (PERMANOVA)

* ANOVA on all pairwise beta-diversity values (_not_ nMDS data points)

>* Same assumptions as ANOVA
* Allows complex models

## Statistically assess beta-diversity

**There was an error in the original slides where the "high" oxygen sample was left in. This is the correct result.**

```{r message=FALSE, warning=FALSE, echo=FALSE}
# Permanova based on low vs. absent oxygen
adonis(OTU.norm[-1,] ~ nmds.data$O2_group[-1], method="bray")
```

## {.flexbox .vcenter}

<font size="8">
Which microbes are causing this difference?
</font>

## Kruskal-Wallis of abundant OTUs

```{r message=FALSE, echo=FALSE, warning=FALSE, results=FALSE}
# From SIMPER (see https://github.com/asteinberger9/seq_scripts for more details)
OTU.list <-c("Otu0001","Otu0002","Otu0003","Otu0006","Otu0004","Otu0005","Otu0010","Otu0011","Otu0009","Otu0014","Otu0016","Otu0013","Otu0017","Otu0015","Otu0018","Otu0025")

# P-values from kruskal.test() of the above OTUs
p.list <- c(0.2752,0.0463,0.04953,0.2752,0.5127,0.5127,0.04953,0.04953,0.5127,0.5127,0.04953,0.2752,0.04953,0.2752,0.04953,0.04311)

# FDR correct p-values
p.list.adjust <- p.adjust(p.list, "fdr")
```

```{r echo=FALSE, warning=FALSE, message=FALSE, fig.height=5}
kw.data <-
  # Extract taxonomy data
  as.data.frame(tax_table(phy.perc)) %>% 
  # Move rowname to variable
  rownames_to_column() %>% 
  dplyr::rename(OTU=rowname) %>% 
  # Merge with kruskal-wallis results
  right_join(data.frame(OTU = OTU.list,
                        pval = p.list,
                        p.adj = p.list.adjust
))


# Abundance plot of selected OTUs
kw.plot <- 
  # Merge kruskal-wallis results with abundance data
  phy.perc %>% 
    # Extract OTUs of interest
    prune_taxa(taxa=OTU.list) %>% 
    # Extract OTU table
    otu_table() %>% 
    as.data.frame() %>% 
    # Move rownames
    rownames_to_column() %>% 
    # Add oxygen level
    mutate(O2_group = nmds.data$O2_group) %>% 
    # Gather OTUs
    gather(-rowname, -O2_group, key=OTU, value=abundance) %>% 
    # Combine with kruskal-wallis results
    full_join(kw.data) %>% 
  # Remove 10 m sample with very high oxygen
  filter(rowname != "Saanich_010") %>% 
  
# Boxplot of relative abundance vs. oxygen
ggplot(aes(x=OTU, y=abundance)) + 
  geom_boxplot(aes(fill=O2_group)) + 
  #  Beautify
  theme_classic() +
  theme(text = element_text(size=18), 
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position="bottom") +
  labs(y="Percent relative abundance", x="")

# Add significance notations
kw.plot +
  annotate("text", x = c(2,3,8,9,13,14,15,16), y = c(42,20,8,8,8,8,8,8), label="*", size=10)
```

## Kruskal-Wallis with FDR correction

```{r echo=FALSE, warning=FALSE, message=FALSE, fig.height=5}
# Change labels to not signif
kw.plot +
  annotate("text", x = c(2,3,8,9,13,14,15,16), y = c(42,20,8,8,8,8,8,8), label="NS", size=5)
```

## Differentially expressed OTUs (DESeq2)
```{r echo=FALSE, message=FALSE, warning=FALSE}
## USE UNNORMALIZED DATA ##
# Remove high O2 sample
phy.subset <- subset_samples(physeq, O2_group != "high")

# Convert to DESeq2 object
O2.deseq <- phyloseq_to_deseq2(phy.subset, ~O2_group)
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
## DESeq2 calculations ##

# Calculate geometric means of OTUs
gm.mean = function(x, na.rm=TRUE){
                 exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
                                 }

geoMeans <- apply(counts(O2.deseq), 1, gm.mean)

# Estimate size factors
O2.deseq <- estimateSizeFactors(O2.deseq, geoMeans = geoMeans)
O2.deseq <- DESeq(O2.deseq, fitType="local")

# Estimate and order p-values
res <- results(O2.deseq)
res <- res[order(res$padj, na.last=NA), ]

# Filter to P < 0.01
sigtab <- res[(res$padj < 0.01), ]
```

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.height=5.5}
# Bind taxonomic data to DE data
sigtab = cbind(as.data.frame(sigtab), as.matrix(tax_table(phy.subset)[rownames(sigtab), ]))

# Rename long taxa
sigtab$Phylum <- revalue(sigtab$Phylum, c("Marinimicrobia_(SAR406_clade)" = "Marinimicrobia"))
sigtab$Genus<- revalue(sigtab$Genus, c("Marinimicrobia_(SAR406_clade)_ge" = "Marinimicrobia"))

# Define genus (y-axis) order
## There are WAY better ways to do this but in the interests of getting the exact plot I wanted, I brute forced the order.
genus.order <- 
#Bact
c("NS9_marine_group_ge","Marinoscillum","NS4_marine_group","NS2b_marine_group",
#Chloro
"SAR202_clade_ge",
#Marin
"Marinimicrobia",
#PAU
"PAUC34f_ge",
#Proteo
"Halieaceae_unclassified","Balneatrix","Pseudohongiella","Emcibacter","JL-ETNP-Y6_ge","PS1_clade_ge","SUP05_cluster_unclassified","Litoreibacter","ZD0417_marine_group","SAR92_clade","OM75_clade","AEGEAN-169_marine_group","Rhodobacteraceae_unclassified","Betaproteobacteria_unclassified","Oceanospirillales_unclassified","Proteobacteria_unclassified","Oligoflexaceae_ge","Milano-WF1B-03_ge","Gammaproteobacteria_unclassified","Alphaproteobacteria_unclassified","Bacteriovoracaceae_unclassified","Peredibacter",
#SBR
"SBR1093_ge",
#Thau
"Candidatus_Nitrosopelagicus","Candidatus_Nitrosopumilus",
#Verruco
"OPB35_soil_group_ge","Opitutae_unclassified","MB11C04_marine_group_ge")

# Plot
deseq.plot <- ggplot(sigtab, aes(y=Genus, x=log2FoldChange, color=Phylum)) + 
  geom_vline(xintercept = 0.0, color = "gray", size = 0.5) +
  geom_point(size=5) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  theme_bw() +
  scale_y_discrete(limits=genus.order) +
  scale_x_continuous(limits=c(-10,10)) +
  theme(text = element_text(size=18))

deseq.plot
```

## Differentially expressed OTUs (DESeq2)
```{r echo=FALSE, message=FALSE, warning=FALSE, fig.height=5.5}
#Highlight OTUs for more investigation
deseq.plot +
  annotate("rect", xmin = -10, xmax = 10, ymin = 6.5, ymax = 8.5,
  alpha = 0.5)
```

## Further investigation

```{r echo=FALSE, message=FALSE, warning=FALSE}
highlight.data <- phy.perc %>% 
  # Extract OTUs of interest
  prune_taxa(taxa=c("Otu0283","Otu0191")) %>% 
  # Extract OTU table
  otu_table() %>% 
  as.data.frame() %>% 
  # Move column names
  rownames_to_column() %>% 
  # Add oxygen levels
  mutate(O2_group = sample_data(phy.perc)$O2_group) %>% 
  # Gather OTUs
  gather(-rowname, -O2_group, key=OTU, value=abundance)
  

highlight.data %>% 
  #Remove high oxygen sample
  filter(rowname != "Saanich_010") %>% 
  
ggplot(aes(x=OTU, y=abundance)) + 
  geom_boxplot(aes(fill=O2_group)) + 
  theme_classic() +
  theme(text = element_text(size=18),
        legend.position="bottom") +
  labs(y="Percent relative abundance", x="") +
  scale_x_discrete(labels=c("PAUC34f", "Halieaceae"))
```

## Deeper into beta-diversity

* Indicator species analysis

* Co-occurrence networks

* Machine learning models like random forest

* Etc...

## On-going challenges

>* More sequencing isn't always better

>* What is a species?

>* Normalization methods

>* Biological relevance

```{r echo=FALSE, eval=FALSE}
## Outline
15min: Intro microbiome and sequencing
20min: Challenges and some solutions
10min: Data intro
10min: nMDS
5min: permanova
5min: DE
5min: Continuing challenges and future research
10min: questions
```